### ISSUE TITLE
Migration from SQLite to MySQL/MariaDB

### ISSUE THREAD

**Issue by Bahama03** [louislam/uptime-kuma#5280]
### 📑 I have found these related issues/pull requests

Related to #5274, wanting to migrate to MySQL.

### 🏷️ Feature Request Type

Deployment

### 🔖 Feature description

With a correct db-config.json file pointing to the MySQL instance, the tables are created but it is treated as a new installation. No tables are migrated.
Tested this with a copy of the database and a separate container.

### ✔️ Solution

In my case I would like to migrate my current installation on the latest version of v1 to v2, but directly migrating/importing in MySQL instead of sqlite3.

**Comment by CommanderStorm** [comment]
Please see https://github.com/louislam/uptime-kuma/wiki/Migration-From-v1-To-v2

> **Can I migrate my existing SQLite database to MariaDB?**
>
> Cannot be done directly. You will need to export your data from SQLite and import it into MariaDB using 3rd party tools.

TL;DR:
We don't currently have the (reviewer, support, triage, testing, ..) resources to build out the interop-tooling (incl. neessesary testing) surrounding this.

Not having this be part of the process is likely fine. v1 is quite limited in its scalablity (see #4500) anyways.
=> If you need the level of monitors that was previously not possible, you are going to be importing using the python wrapper / .. anyway. 🤷🏻‍♂️

**Comment by derekoharrow** [comment]
It would at least be good to have a guide on how to export, convert and import from sqlite to MySQL/MariaDB. I tried this manually myself yesterday and couldn't get it to work.

**Comment by monster010** [comment]
It is possible to migrate, I did it manually and by manual I mean:

- Create new Kuma instance (new folder, complete new installation)
- Run installer (did not use embedded mariadb because embedded caused problems)
- Opening kuma.db and manually transferring the data, making adjustments here and there because otherwise it doesn't work
  - For example, a dump from a SQLite contains `''` instead of `NULL`. `AUTO_INCREMENT` is `AUTOINCREMENT` in SQLite and things like that.

I'm running MariaDB now and haven't noticed any problems so far.

**Comment by Apashh** [comment]
@monster010 nice !! It will be interesting to note this in Wiki migrate v1 SQLite to v2 Mariadb.

**Comment by CommanderStorm** [comment]
Sorry, but as stated above:
We don't have the resources to support this part. We still don't have the resources to maintain this part, even if it were to live in the wiki.

If you create a gist with an update script, that is fine for us though, we can link that.

**Comment by Mazvy** [comment]
If someone, even if unofficially, can shed some light into the steps required, or perhaps prepared a script in order to automate this - please share.

**Comment by cmbcbe** [comment]
no success while importing converted db to mariadb

```
apt instal python3 git
git clone https://github.com/majidalavizadeh/sqlite-to-mysql.git
cd sqlite-to-mysql
python3 export.py /srv/docker/uptime-kuma/data/kuma.db /srv/docker/uptime-kuma/data/kuma_mariadb.sql
```

i will try to play with python sqlite3mysql

**Comment by cmbcbe** [comment]
Hello, i have have successfully migrate from SQlite3 to Mariadb in this way on V2 beta 1.
First of all, if you don't want that the migration take a long time, pause all your monitors then clear statistics data, the idea behind is to get back ours monitors only.

Than you need to shutdown uptime-kuma container to get the sqlite3 database in a concice state (kuma.db)

Then modify your docker compose to add mariadb in the game, here is the example of mine that i use in a swarm docker environnement

```
version: '3.8'

services:
  webui:
    image: louislam/uptime-kuma:nightly2
    deploy:
      endpoint_mode: dnsrr
      replicas: 1
      restart_policy:
        condition: any
      placement:
        constraints:
          - node.hostname == docker003
    volumes:
      - /srv/docker/uptime-kuma/data:/app/data
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - UPTIME_KUMA_DISABLE_FRAME_SAMEORIGIN=1
    dns:
      - 10.74.1.11
      - 10.75.3.15
    networks:
      - uptime-kuma_net
      - npm

  mariadb:
    image: mariadb:11.4
    deploy:
      endpoint_mode: dnsrr
      replicas: 1
      restart_policy:
        condition: any
      placement:
        constraints:
          - node.hostname == docker003
    environment:
      - MYSQL_ROOT_PASSWORD=Root@ccess1234!
      - MYSQL_DATABASE=uptime_kuma
      - MYSQL_USER=uptime_kuma
      - MYSQL_PASSWORD=Kum@ccess1234!
    volumes:
      - /srv/docker/uptime-kuma/mariadb/data:/var/lib/mysql
      - /srv/docker/uptime-kuma/mariadb/my.cnf:/etc/mysql/conf.d/my.cnf
    networks:
      - uptime-kuma_net
      - npm

networks:
  npm:
     external: true
  uptime-kuma_net:
    driver: overlay
    internal: true
    attachable: false
```

Thus update the db config file in uptime db-config.json to point to mariadb
example:
{
    "type": "mariadb",
    "port": 3306,
    "hostname": "uptime-kuma_mariadb",
    "username": "uptime_kuma",
    "password": "mypassword",
    "dbName": "uptime_kuma"
}

Spinup Uptime-Kuma container thus it will populate the MariadB database schema for Uptime.
I have export the empty database [here if needed](https://files.cmbc.be/api/public/dl/X8bQSnhC/sources/Linux/docker/uptime-kuma/uptime_kuma_v2.sql)
After the database creation, shutdown uptime container to avoid conflict while importing data and live production.

In the way you want, the idea is to use a tool called sqlite3mysql to import kuma.db to your mariadb, here is how i've done:

i have copy from the docker host the file kuma.db inside uptime-kuma container data directory to the data directory of mariadb (corresponding inside the container to /var/lib/mysql), then while the mariadb container is running, i enter inside and do this:

`docker exec -it <your container id of mariadb:11.4> /bin/bash`
`apt update && apt install python3-pip`
`pip install sqlite3-to-mysql --break-system-packages`
`sqlite3mysql -f /var/lib/mysql/kuma.db -d uptime_kuma -u root -p -h 127.0.0.1 -K -i IGNORE`
The option "ignore" is used because some table time field does not work from sqlite3 to mariadb (format incompatible), because we have drop the statistics this not a problems.

you will get a screen like that while migrating, it can take a while
![Image](https://github.com/user-attachments/assets/89178eb0-5f15-4ce8-9b2e-9458d64d0e58)

Start your uptime-kuma container and you should retrive your data on screen coming from mariadb!

**Comment by Bahama03** [comment]
@cmbcbe Thanks a lot! Works as a charm!

**Comment by lauer** [comment]
@cmbcbe  Thanks for sharing - did you clean statistik before the migration to MariaDB? and how large was your sqlite db?

**Comment by ctoma** [comment]
For anyone reading this later I successfully migrated a non-docker install to the v2 beta with no data loss (all history intact). Here's how I did it.

1. Install MariaDB on the machine
2. Stop your old uptime kuma so all the data remains stable.
3. `git clone https://github.com/louislam/uptime-kuma.git uptime-kuma-v2`
4. `cd uptime-kuma-v2`
5. `npm ci --omit dev && npm run download-dist`
6. `node server/server.js`
7. Open up the web browser and enter in the MariaDB connection details, you should see the database tables created and then an empty Uptime Kuma install. The goal is to have all the tables created and later the sqlite3mysql script will fill the tables (it will not create the tables because of the -K flag which is necessary).
8. Stop your new uptime kuma instance
9. Now copy over your `kuma.db` file from your old uptime kuma data directory into your new uptime kuma data directory
10. Inside of your data directory rename `db-config.json` to `db-config.json.bak` (or whatever you want), we will need it later
11. `node server/server.js`
12. Now it will detect your `kuma.db` sqlite database and migrate all of your old data into the new format, but it will keep everything in sqlite. For reference my `kuma.db` file was about 1GB, 350 monitors, 90 day history. It took about an hour to complete on a `t3.small`, and it's worth noting that the progress indicator is bugged and showed it complete when it was only about 60% of the way through. So just be patient and it will work.
13. After all the data is migrated stop the server.
14. Run the commands that @cmbcbe mentioned using `sqlite3mysql` which fills the database. This took about 1 minute on the `t3.small`
15. Delete `db-config.json` (which contains your sqlite details), and rename your `db-config.json.bak` to `db-config.json`.
16. Launch Uptime Kuma and you should now be on v2 with all data migrated to MariaDB and no data loss.

**Comment by rsubr** [comment]
Thanks for the steps, this still works on beta4.

> `pip install sqlite3-to-mysql --break-system-packages` `sqlite3mysql -f /var/lib/mysql/kuma.db -d uptime_kuma -u root -p -h 127.0.0.1 -K -i IGNORE`

I also had to `pip3 install typing_extensions --break-system-packages`, and then I was able to run `sqlite3mysql`.

I used the embedded mariadb option and my migration command looked like this:

```
sqlite3mysql -f kuma.db -k /app/data/run/mariadb.sock -d kuma -u root -K -i IGNORE
```

**Comment by JuanPabloToniolo-Udesa** [comment]
> Thanks for the steps, this still works on beta4.
> 
> > `pip install sqlite3-to-mysql --break-system-packages` `sqlite3mysql -f /var/lib/mysql/kuma.db -d uptime_kuma -u root -p -h 127.0.0.1 -K -i IGNORE`
> 
> I also had to `pip3 install typing_extensions --break-system-packages`, and then I was able to run `sqlite3mysql`.
> 
> I used the embedded mariadb option and my migration command looked like this:
> 
> ```
> sqlite3mysql -f kuma.db -k /app/data/run/mariadb.sock -d kuma -u root -K -i IGNORE
> ```

Also running now with a db of 2.5GB with 250 days worth of data, more than 20 monitors and it still working on I don't count that it will took so much processing so it's taking too long on lightsail with 2 cores and 2gb ram. Will report if finish 😅

Also already running beta4


UPDATE: It finished ok but statistics not reflect the older heartbeats.

UPDATE2: Running the migration again on external computer got it working perfect. in less than 10 minutes

**Comment by badnetmask** [comment]
> Thanks for the steps, this still works on beta4.
> 
> > `pip install sqlite3-to-mysql --break-system-packages` `sqlite3mysql -f /var/lib/mysql/kuma.db -d uptime_kuma -u root -p -h 127.0.0.1 -K -i IGNORE`
> 
> I also had to `pip3 install typing_extensions --break-system-packages`, and then I was able to run `sqlite3mysql`.
> 
> I used the embedded mariadb option and my migration command looked like this:
> 
> ```
> sqlite3mysql -f kuma.db -k /app/data/run/mariadb.sock -d kuma -u root -K -i IGNORE
> ```

Life saver! This worked perfectly! Thank you!

**Comment by SchiWaGoA78** [comment]
@cedric-marcoux thx a lot for this charming tutorial.
i migrated sqlite to mariadb on Version 2.0.2 docker and run instant.
the performance is much better. 💯 👍 🥇

**Comment by powerslid3** [comment]
@cedric-marcoux @ctoma thank you very much, successfully migrated from SQLite to MariaDB 12.1.2 and Uptime-Kuma 2.0.2
Performance is significantly better

**Comment by Alien-lv426** [comment]
@ctoma I wanted to add my thanks for your instructions. Managed to migrate from sqlite to mariadb with everything I wanted intact.

**Comment by lluisd** [comment]
thank you! it worked also for me with uptime-kuma 2.2.1
